lib/client/keystore.go

```diff
@@ -378,33 +373,61 @@ type withKubeCerts struct {
 }
 
 // TODO(awly): unit test this.
-func (o withKubeCerts) getKey(dirPath, username string, key *Key) error {
-	kubeDir := filepath.Join(dirPath, username+kubeDirSuffix, o.teleportClusterName)
-	kubeFiles, err := ioutil.ReadDir(kubeDir)
-	if err != nil && !os.IsNotExist(err) {
-		return trace.Wrap(err)
-	}
-	if key.KubeTLSCerts == nil {
-		key.KubeTLSCerts = make(map[string][]byte)
-	}
-	for _, fi := range kubeFiles {
-		data, err := ioutil.ReadFile(filepath.Join(kubeDir, fi.Name()))
-		if err != nil {
+func (o withKubeCerts) getKey(store LocalKeyStore, idx keyIndex, key *Key) error {
+	switch s := store.(type) {
+	case *FSLocalKeyStore:
+		dirPath := s.dirFor(idx.proxyHost)
+		kubeDir := filepath.Join(dirPath, idx.username+kubeDirSuffix, o.teleportClusterName)
+		kubeFiles, err := ioutil.ReadDir(kubeDir)
+		if err != nil && !os.IsNotExist(err) {
 			return trace.Wrap(err)
 		}
-		kubeCluster := strings.TrimSuffix(filepath.Base(fi.Name()), fileExtTLSCert)
-		key.KubeTLSCerts[kubeCluster] = data
+		if key.KubeTLSCerts == nil {
+			key.KubeTLSCerts = make(map[string][]byte)
+		}
+		for _, fi := range kubeFiles {
+			data, err := ioutil.ReadFile(filepath.Join(kubeDir, fi.Name()))
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			kubeCluster := strings.TrimSuffix(filepath.Base(fi.Name()), fileExtTLSCert)
+			key.KubeTLSCerts[kubeCluster] = data
+		}
+	case *MemLocalKeyStore:
+		if len(key.KubeTLSCerts) == 0 {
+			idx.clusterName = o.teleportClusterName
+			stored, ok := s.inMem[idx]
+			if !ok {
+				return trace.NotFound("key for %v not found", idx)
+			}
+			key.KubeTLSCerts = stored.KubeTLSCerts
+		}
+	default:
+		return trace.BadParameter("unexpected key store type %T", store)
 	}
 	if key.ClusterName == "" {
 		key.ClusterName = o.teleportClusterName
 	}
 	return nil
 }
 
-func (o withKubeCerts) deleteKey(dirPath, username string) error {
-	kubeCertsDir := filepath.Join(dirPath, username+kubeDirSuffix, o.teleportClusterName)
-	if err := os.RemoveAll(kubeCertsDir); err != nil {
-		return trace.Wrap(err)
+func (o withKubeCerts) deleteKey(store LocalKeyStore, idx keyIndex) error {
+	switch s := store.(type) {
+	case *FSLocalKeyStore:
+		dirPath := s.dirFor(idx.proxyHost)
+		kubeCertsDir := filepath.Join(dirPath, idx.username+kubeDirSuffix, o.teleportClusterName)
+		if err := os.RemoveAll(kubeCertsDir); err != nil {
+			return trace.Wrap(err)
+		}
+	case *MemLocalKeyStore:
+		idx.clusterName = o.teleportClusterName
+		stored, ok := s.inMem[idx]
+		if !ok {
+			return trace.NotFound("key for %v not found", idx)
+		}
```

Comment:
this is not an error for `FSLocalKeyStore` (`os.RemoveAll` doesn't return an error if the path is not found), so let's not make this an error here either.